File: dcepc206.py
#!/usr/bin/env python3
# http://www.spoj.com/problems/DCEPC206/
# It's a Murder!

# Once detective Saikat was solving a murder case. While going to the crime
# scene he took the stairs and saw that a number is written on every stair. He
# found it suspicious and decides to remember all the numbers that he has seen
# till now. While remembering the numbers he found that he can find some pattern
# in those numbers. So he decides that for each number on the stairs he will note
# down the sum of all the numbers previously seen on the stairs which are smaller
# than the present number. Calculate the sum of all the numbers written on his
# notes diary.

# Input

# First line gives T, number of test cases.
# 2T lines follow.
# First line gives you the number of stairs N
# Next line gives you N numbers written on the stairs.

# Output

# For each test case output one line giving the final sum for each test case.

# Constraints

# T<=10
# 1<=N<=10^5

# All numbers will be between 0 and 10^6.

class BinaryTree:

    def __init__(self, v):
        self.left = None
        self.right = None
        self.parent = None
        self.value = v
        self.replicas = 1
        self._sum = 0

    def __repr__(self):
        return str(self)

    def __str__(self):
        return "[{} {} {}] ({}) ({})".format(self.value, self._sum, self.sum, self.left, self.right)

    @property
    def sum(self):
        total = self._sum
        current = self
        while True:
            if current.parent is None:
                return total
            parent = current.parent
            if parent.value < self.value:
                total += (parent.value*parent.replicas) + parent._sum
            current = current.parent

    def insert(self, v):
        current = self
        s = 0
        while True:
            if current.value == v:
                current.replicas += 1
                return current._sum + s
            elif current.value < v:
                s += (current.value*current.replicas) + current._sum
                if current.right is None:
                    node = BinaryTree(v)
                    node.parent = current
                    current.right = node
                    return s
                else:
                    current = current.right
            else:
                current._sum += v
                if current.left is None:
                    node = BinaryTree(v)
                    node.parent = current
                    current.left = node
                    return s
                else:
                    current = current.left


def solve(stair):
    tree = BinaryTree(stair.pop(0))
    total = 0
    for step in stair:
        total += tree.insert(step)
    return total


def main():
    t = int(input())
    for i in range(t):
        size = int(input())
        stair = [int(n) for n in input().split(' ')[:size]]
        print(solve(stair))


if __name__ == "__main__":
    main()

File: dcepc206.c
#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>

#define MAX_STEPS 100000

struct node {
    struct node *left;
    struct node *right;
    struct node *parent;
    int64_t value;
    int64_t replicas;
    int64_t sum;
    int height;
};

struct tree {
    struct node *root;
    struct node nodes[MAX_STEPS];
    int64_t pointer;
};

int clean_tree(struct tree *t) {
    t->pointer = 0;
}

struct node* alloc_node(struct tree *t) {
    return &(t->nodes[t->pointer++]);
}

struct node* mknode(struct tree *t, struct node *current, int64_t v) {
    struct node *n = alloc_node(t); \
    n->left = NULL;
    n->right = NULL;
    n->parent = current;
    n->value = v;
    n->sum = 0;
    n->replicas = 1;
    n->height = 1;
    return n;
} 

void rotate_left(struct tree* t, struct node* n) {
    struct node **root;
    if (n->parent) {
        if (n->parent->right == n) {
            root = &n->parent->right;
        } else {
            root = &n->parent->left;
        }
    } else {
        root = &t->root;
    }
    assert(n->right);
    *root = n->right;
    (*root)->parent = n->parent;
    n->right = (*root)->left;
    if (n->right) {
        n->right->parent = n;
    }
    (*root)->left = n;
    n->parent = *root;
    (*root)->sum += (n->value*n->replicas) + n->sum;

    n->height = max_height(n->left, n->right) + 1;
    (*root)->height = max_height((*root)->left, (*root)->right) + 1;
}

void rotate_right(struct tree* t, struct node* n) {
    struct node **root;
    if (n->parent) {
        if (n->parent->right == n) {
            root = &n->parent->right;
        } else {
            root = &n->parent->left;
        }
    } else {
        root = &t->root;
    }
    assert(n->left);
    n->left->parent = n->parent;
    *root = n->left;
    n->left = n->left->right;
    if (n->left) {
        n->left->parent = n;
    }
    assert(n->sum >= ((*root)->value * (*root)->replicas) - (*root)->sum);
    n->sum = n->sum - ((*root)->value * (*root)->replicas) - (*root)->sum;
    (*root)->right = n;
    n->parent = *root;

    n->height = max_height(n->left, n->right) + 1;
    (*root)->height = max_height((*root)->left, (*root)->right) + 1;
}

int get_height(struct node *n) {
    if (!n) {
        return 0;
    } else {
        return n->height;
    }
}

int max_height(struct node *a, struct node *b) {
    int ha = get_height(a);
    int hb = get_height(b);
    return ha > hb ? ha : hb;
}

void balance(struct tree* t, struct node* n, int64_t value) {
    int diff;
    while (n) {
        diff = get_height(n->left) - get_height(n->right);

        if (diff < -1 && value > n->right->value) {
            rotate_left(t, n);
        } else if (diff > 1  && value < n->left->value) {
            rotate_right(t, n);
        } else if (diff < -1 && value < n->right->value) {
            rotate_right(t, n->right);
            rotate_left(t, n);
        } else if (diff > 1 && value > n->left->value) {
            rotate_left(t, n->left);
            rotate_right(t, n);
        }

        if (!n->left && !n->right) {
            n->height = 1;
        } else {
            n->height = max_height(n->left, n->right) + 1;
        }
        n = n->parent;
    }
}

int64_t insert(struct tree* t, int64_t value) {
    int p = 0;
    struct node *current = t->root;
    int64_t s = 0;
    while (1) {
        if (current->value == value) {
            current->replicas++;
            return current->sum + s;
        } else if (current->value < value) {
            s += current->value*current->replicas + current->sum;
            if (!current->right) {
                current->right = mknode(t, current, value);
                balance(t, current, value);
                return s;
            } else {
                current = current->right;
            }
        } else {
            current->sum += value;
            if (!current->left) {
                current->left = mknode(t, current, value);
                balance(t, current, value);
                return s;
            } else {
                current = current->left;
            }
        }
    }
}

int64_t solve(struct tree *t, int64_t stairs[], int size) {
    int i;
    int64_t total = 0;
    clean_tree(t);
    t->root = mknode(t, NULL, stairs[0]);

    for (i = 1; i < size; i++) {
        total += insert(t, stairs[i]);
    }

    return total;
}

int main(void) {
    int i, t;
    int tests;
    int steps;
    int64_t stairs[MAX_STEPS];
    struct tree tree;

    scanf("%d", &tests);
    for (t = 0; t < tests; t++) {
        scanf("%d", &steps);
        for (i = 0; i < steps; i++) {
            scanf("%" PRIu64, &stairs[i]);
        }
        printf("%" PRIu64 "\n", solve(&tree, stairs, steps));
    }
}
